luarules/gadgets/unit_custom_weapons_cluster.lua

```diff
@@ -59,281 +58,258 @@ local cos   = math.cos
+local spGetUnitRadius         = Spring.GetUnitRadius
+local spGetUnitsInSphere      = Spring.GetUnitsInSphere
+local spSpawnProjectile       = Spring.SpawnProjectile
 
-local GAME_SPEED         = Game.gameSpeed
-local mapGravity         = Game.gravity / GAME_SPEED / GAME_SPEED * -1
+local gameSpeed  = Game.gameSpeed
+local mapGravity = Game.gravity / (gameSpeed * gameSpeed) * -1
 
-local SetWatchExplosion  = Script.SetWatchExplosion
+--------------------------------------------------------------------------------
+-- Initialize ------------------------------------------------------------------
 
---------------------------------------------------------------------------------------------------------------
--- Initialize ------------------------------------------------------------------------------------------------
-
--- Information table for cluster weapons
+defaultSpawnTtl = defaultSpawnTtl * gameSpeed
 
 local spawnableTypes = {
-    Cannon          = true  ,
-    EMGCannon       = true  ,
-    Fire            = false , -- but possible
-    LightningCannon = false , -- but possible
-    MissileLauncher = false , -- but possible
+	Cannon          = true  ,
+	EMGCannon       = true  ,
 }
 
-local dataTable      = {} -- Info on each cluster weapon
-local wDefNamesToIDs = {} -- Says it on the tin
-
-for wdid, wdef in pairs(WeaponDefs) do
-    wDefNamesToIDs[wdef.name] = wdid
-
-    if wdef.customParams and wdef.customParams[customParamName] then
-        dataTable[wdid] = {}
-        dataTable[wdid].number  = tonumber(wdef.customParams.number) or defaultSpawnNum
-        dataTable[wdid].def     = wdef.customParams.def
-        dataTable[wdid].projDef = -1
-        dataTable[wdid].projTtl = defaultSpawnTtl
-        dataTable[wdid].projVel = defaultVelocity / GAME_SPEED
-
-        -- Enforce limits, eg the projectile count, at init.
-        dataTable[wdid].number  = min(dataTable[wdid].number, maxSpawnNumber)
-
-        -- When the cluster munition name isn't specified, search for the default.
-        if dataTable[wdid].def == nil then
-            local search = ''
-            for word in string.gmatch(wdef.name, '([^_]+)') do
-                search = search == '' and word or search .. '_' .. word
-                if UnitDefNames[search] ~= nil then
-                    dataTable[wdid].def = search .. '_' .. defaultSpawnDef
-                end
-            end
-            -- There's still the chance we haven't found anything, so:
-            if dataTable[wdid].def == nil then
-                Spring.Echo('[clustermun] [warn] Did not find cluster munition for weapon id ' .. wdid)
-                dataTable[wdid] = nil
-            end
-        end
-    end
+local clusterWeaponDefs = {}
+
+for unitDefName, unitDef in pairs(UnitDefNames) do
+	for _, weapon in pairs(unitDef.weapons) do
+		local weaponDefID, weaponDef = weapon.weaponDef, WeaponDefs[weapon.weaponDef]
+		local custom = weaponDef.customParams
+
+		if custom.cluster then
```

Comment:
Why have a separate `cluster` and not just look for the presence of `cluster_def` which AFAICT is mandatory anyway?

Reply:
Because custom params don't support tables, but in our hearts we're using them. If we close our eyes and wish hard enough, we can see it.
```lua
cluster = {
  def = 'cluster_weapon',
  number = 10,
}
```

Reply:
But all it does in practice is make the perfectly valid-looking
```lua
cluster_def = "bla",
cluster_number = 123,
```
not work and pollute the API with the extra redundant bool. Worse yet, if table support gets added to customparams the `cluster` entry will already be a bool in various places (including harder-to-find stuff like tools that gather unit stats for website etc) that will make migrating to a table more annoying. If you insist on having a bool why not `is_cluster` or `cluster_enabled` or such?